src/core/search/block_list.cc

```diff
@@ -0,0 +1,128 @@
+#include "core/search/block_list.h"
+
+namespace dfly::search {
+
+using namespace std;
+
+template <typename C> bool BlockList<C>::Insert(DocId t) {
+  auto block = FindBlock(t);
+  if (block == blocks_.end())
+    block = blocks_.insert(blocks_.end(), C{blocks_.get_allocator().resource()});
+
+  if (!block->Insert(t))
+    return false;
+
+  size_++;
+  TrySplit(block);
+  return true;
+}
+
+template <typename C> bool BlockList<C>::Remove(DocId t) {
+  if (auto block = FindBlock(t); block != blocks_.end() && block->Remove(t)) {
+    size_--;
+    TryMerge(block);
+    return true;
+  }
+
+  return false;
+}
+
+template <typename C> typename BlockList<C>::BlockIt BlockList<C>::FindBlock(DocId t) {
+  DCHECK(blocks_.empty() || blocks_.back().Size() > 0u);
+
+  if (!blocks_.empty() && t >= *blocks_.back().begin())
+    return --blocks_.end();
+
+  // Find first block that can't contain t
+  auto it = std::upper_bound(blocks_.begin(), blocks_.end(), t,
+                             [](DocId t, const C& l) { return *l.begin() > t; });
+
+  // Move to previous if possible
+  if (it != blocks_.begin() && blocks_.size() > 0)
```

Comment:
I think you don't need the check blocks_.size() > 0

Reply:
True, because for an empty vector begin() = end()